org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -601,10 +603,36 @@ private static boolean isJavaIdentifierOrPeriod(char ch) {
 		return Character.isJavaIdentifierPart(ch) || ch == '.';
 	}
 
+	public static boolean isFolder(String uriString) {

```

Comment:
return findFolder(uriString).exists();

Reply:
`findFolder(uriString)` always return a new IResource object of type IContainer (it just creates it) even if `uriString` represents a real file (not a folder) as well as not existing file/folder (exactly as well as `findFile(ueiString)` returns IFile even if `uriString` represents a real folder).  So, `return findFolder(uriString).exists()` will save us from not existing folder, but still return true in case of synched files - which is wrong. Also it will return false in case of not-synched, but existing folders - which is also wrong.

My goal is to report to jdt.ls on every new just created folder (or a package) in order to make it go through `ProjectsManager.fileChanged(String, CHANGE_TYPE)` until it refreshes the specified IResource: https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java#L275 (in case the resource is folder) which is impossible (or involves some other problems/errors) with current realization.
If I do not report folder changes (create/delete/update) to jdt.ls and as such those folders stay not-synchronized (they exist in file system, but not in workspace) then such folders/packages cannot be used as a destination target package of copy/move operations (they just stay invisible for the workspace).

I know that I do an extra refresh of resource in `isFolder()`, but otherwise I don't know a 100%-clear way to detect if a specified uri represents a folder or a file.

Reply:
>>   findFolder(uriString) always return a new IResource object of type IContainer (it just creates it)

> findFolder(uriString) doesn't create it.

`JDTUtils.findFolder(...)` creates a new resource of type IContainer as well as `JDTUtils.findFile(...) ` creates a new resource object of type IFile, independently of real resource type represented by the given URI (either it points to a file or a folder), because the execution of both methods goes through `org.eclipse.core.internal.resources.Workspace.newResource(IPath, int)` where new resource object of specified type is created for the given path.

In your test case you're using 'IFile.create(..)` and `project.refreshLocal(DEPTH_INFINITE)` - the methods which create a file in workspace or even refresh the workspace project file structure. This make all the file system resources to became "visible" for workspace and synchronized before they are tested with `isFolder()`, `findFile()` and `findFolder()`.
In my case, the folders to be reported are not synchronized with workspace - I do actually reporting them to jdt.ls in order to synch them in workspace.

And I don't actually care where exactly `refresh()`is called - in `ProjectManager.fileChanged()` or a bit earlier in `JDTUtils.isFolder()` method - both places suite my needs.

I'll look at your test more deep later - maybe I'm missing something...

Reply:
> org.eclipse.core.internal.resources.Workspace.newResource(IPath, int) where new resource object of specified type is created for the given path.

See javadoc for org.eclipse.core.resources.IContainer.getFile(IPath) and org.eclipse.core.internal.resources.Container.getFile(IPath)

```
	@Test
	public void testFindFileFindFolder() throws Exception {
		IProject project = WorkspaceHelper.getProject(ProjectsManager.DEFAULT_PROJECT_NAME);
		IFolder org = project.getFolder("/src/org");
		org.delete(true, null);
		org.create(true, true, null);
		IFolder iFolder = project.getFolder("/src/org/eclipse");
		URI uriFolder = iFolder.getLocationURI();
		IFile file = JDTUtils.findFile(uriFolder.toString());
		assertFalse(file.exists());
		IContainer folder = JDTUtils.findFolder(uriFolder.toString());
		assertFalse(folder.exists());
		project.refreshLocal(IResource.DEPTH_INFINITE, null); // it doesn't help
		assertFalse(file.exists());
		assertFalse(folder.exists());
		iFolder.create(true, false, null); // create the folder
		assertTrue(folder.exists());
		assertTrue(folder.getType() == IResource.FOLDER && folder.exists());
		assertFalse(file.getType() == IResource.FOLDER && file.exists());
		assertTrue(file.getType() == IResource.FILE);
		assertFalse(file.getType() == IResource.FILE && file.exists());
		org.delete(true, null);
	}
```

> And I don't actually care where exactly refresh()is called - in ProjectManager.fileChanged() or a bit earlier in JDTUtils.isFolder() method - both places suite my needs.

You are right. refresh() should be called in isFolder().

Reply:
Two more notes:
- Your test case doesn't follow my case workflow: I cannot change workspace resources (by calling methods like `IFolder.create()`, `IFile.create()`, `IFolder.delete()` and so on). Not acceptable for us.
- It looks like my testcase still has a problem as it relies on folders that were created by some other test case. It's a bug that I must fix.

Reply:
@snjeza I have fixed my testcase. In short the workflow is the following:
1. [from outside of jdt.ls] some folders (packages) are created by using java.io-API
2. [inside jdt.ls] URI of a changed folder is provided to jdt.ls, jdt.ls is to react by refreshing the folder in its workspace (by using org.eclipse.core.resources-API) in `ProjectManager.fileChanged()` method (here's the work of `isFolder()`, `findFolder() and `findFile()` methods is being tested)